Add: An introductory guide for text generation

MKhalusova opened the pull request:
Context: current [text generation doc](https://huggingface.co/docs/transformers/main/en/main_classes/text_generation) is large, difficult to navigate, and may be overwhelming to users who are new to text generation.  

I suggest splitting the document into two parts: 
1. an introductory guide that explains the basics and provides some examples
2. trimmed down API reference to be used for looking up specific parameter descriptions (https://github.com/huggingface/transformers/pull/21112)
 
This PR is the first part. It adds a guide that introduces readers to the text generation strategies, explains the generation configuration defaults, and provides some examples. The second part will follow in a separate PR.

MKhalusova committed:
Part of the "text generation" rework: adding a high-level overview of the text generation strategies

MKhalusova committed:
code samples update via make style

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_21090). All of your documentation changes will be reflected on that endpoint.

MKhalusova committed:
fixed a few formatting issues

sgugger commented on docs/source/en/generation_strategies.mdx:
```diff
@@ -0,0 +1,287 @@
+You can also store several generation parametrizations in a single directory, making use of the `config_file_name`
+argument in `save_pretrained`. You can later instantiate them with `from_pretrained`. This is useful if you want to
+store several generation configurations for a single model (e.g. one for creative text generation with sampling, and
+other for summarization with beam search). You must have the right Hub permissions to add configuration files to a model.
```
That's if you want to push them to the Hub, which is not shown here. So maybe show an example of this? Or remove the sentence otherwise.

MKhalusova replied:
I am not sure I understand. There's a `push_to_hub=True` in the `save_pretrained`. Doesn't this upload the config to Hub?

sgugger replied:
Ah, I missed it! All good, thanks!

sgugger reviewed (commented):
Very nice new doc, thanks a lot for writing it! I've left a couple of comments.

MKhalusova commented on docs/source/en/generation_strategies.mdx:
```diff
@@ -0,0 +1,287 @@
+
+Text generation is essential to many NLP tasks, such as open-ended text generation, summarization,
+translation, and more. It also plays a role in a variety of mixed-modality applications that have text as an output, e.g.
+speech-to-text, vision-to-text, etc. Some of the models that can generate text in an auto-regressive manner include
+GPT2, XLNet, OpenAi-GPT, CTRL, TransformXL, XLM, Bart, and T5. You can find examples of using the `generate()` method to produce
```
@sgugger I could either 
a) Bart and T5 from the sentence, or
b) remove the words "auto-regressive", and then add more models (e.g. speech to text)

Which option would you prefer?

MKhalusova commented on docs/source/en/generation_strategies.mdx:
````diff
@@ -0,0 +1,287 @@
+```
+
+### Contrastive Search
+The Contrastive search decoding strategy was proposed in the 2022 paper <a href="https://arxiv.org/abs/2202.06417">"A Contrastive Framework for Neural Text Generation"</a>.
````
I initially used markdown, but it wasn't rendering for some reason. Maybe the quotation marks were the issue. I'll switch back to markdown, and hopefully, it will render.

MKhalusova commented on docs/source/en/generation_strategies.mdx:
```diff
@@ -0,0 +1,287 @@
+Even if the default decoding strategy mostly works for your task, you can still tweak a few things. Some of the
+commonly adjusted parameters include:
+
+- `max_new_tokens`: the maximum number of tokens to generate. In other words, the size of the output sequence, not
+including the tokens in the prompt.
+- `max_length`: similar to `max_new_tokens`, this parameter limits the size of the generated output, however, it takes
```
I am not sure about it, these seem rather straightforward to understand, plus I am using max_new_tokens in other examples where I need to generate a longer output.

MKhalusova replied:
According to @gante `max_length` will be deprecated, so I will remove it from here. But I will add `num_return_sequences`.

MKhalusova committed:
Apply suggestions from review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

MKhalusova committed:
fixed spaces, and switched two links to markdown

stevhliu commented on docs/source/en/generation_strategies.mdx:
```diff
@@ -0,0 +1,287 @@
+Here, we'll show what parameters control some of the strategies, and illustrate how you can use them.
+
+### Greedy Search
+Greedy search is the default decoding strategy. For generate() method to use it, `num_beams` has to be set to 1, and
```
```suggestion
[`generate`] uses greedy search decoding by default so you don't have to pass any parameters to enable it. This means the parameters `num_beams` is set to 1 and `do_sample=False`.
```

stevhliu commented on docs/source/en/generation_strategies.mdx:
````diff
@@ -0,0 +1,287 @@
+['I look forward to seeing you all again!\n\n\n\n\n\n\n\n\n\n\n']
+```
+
+### Contrastive Search
````
Add a blank line after the header

```suggestion
### Contrastive search
```

stevhliu commented on docs/source/en/generation_strategies.mdx:
````diff
@@ -0,0 +1,287 @@
+```
+
+### Contrastive Search
+The Contrastive search decoding strategy was proposed in the 2022 paper [A Contrastive Framework for Neural Text Generation](https://arxiv.org/abs/2202.06417).
````
```suggestion
The contrastive search decoding strategy was proposed in the 2022 paper [A Contrastive Framework for Neural Text Generation](https://arxiv.org/abs/2202.06417).
```

stevhliu commented on docs/source/en/generation_strategies.mdx:
```diff
@@ -0,0 +1,287 @@
+The Contrastive search decoding strategy was proposed in the 2022 paper [A Contrastive Framework for Neural Text Generation](https://arxiv.org/abs/2202.06417).
+It demonstrates superior results for generating non-repetitive yet coherent long outputs. To learn how contrastive search
+works, check out [this blog post](https://huggingface.co/blog/introducing-csearch).
+The two main parameters that enable and control the behavior of the contrastive search are `penalty_alpha` and `top_k`:
```
It may be nice to add links to the two parameters so users can click on it and learn what they are/do.

```suggestion
The two main parameters that enable and control the behavior of contrastive search are `penalty_alpha` and `top_k`:
```

stevhliu commented on docs/source/en/generation_strategies.mdx:
````diff
@@ -0,0 +1,287 @@
+```
+
+### Beam-search decoding
+As opposed to the greedy search, beam search decoding keeps several hypotheses at each time step and eventually chooses
````
Just to mix things up a little :)

```suggestion
Unlike greedy search, beam-search decoding keeps several hypotheses at each time step and eventually chooses
```

stevhliu commented on docs/source/en/generation_strategies.mdx:
```diff
@@ -0,0 +1,287 @@
+### Beam-search decoding
+As opposed to the greedy search, beam search decoding keeps several hypotheses at each time step and eventually chooses
+the hypothesis that has the overall highest probability for the entire sequence. This has the advantage of identifying high-probability
+sequences that start with a lower probability initial tokens and would've been ignored by the greedy search.
```
```suggestion
sequences that start with lower probability initial tokens and would've been ignored by the greedy search.
```

stevhliu commented on docs/source/en/generation_strategies.mdx:
```diff
@@ -0,0 +1,287 @@
+the hypothesis that has the overall highest probability for the entire sequence. This has the advantage of identifying high-probability
+sequences that start with a lower probability initial tokens and would've been ignored by the greedy search.
+
+To enable this decoding strategy, specify the `num_beams` that is greater than 1.
```
Maybe clarify what exactly the number of beams do, something like:

```suggestion
To enable this decoding strategy, specify `num_beams` greater than 1. The number of beams specified corresponds to how many time-steps to look ahead.
```

stevhliu commented on docs/source/en/generation_strategies.mdx:
```diff
@@ -0,0 +1,287 @@
+
+### Beam-search multinomial sampling
+As the name implies, this decoding strategy combines beam search with multinomial sampling. You need to specify
+the `num_beams` that is greater than 1, and set `do_sample=True` to use this decoding strategy.
```
```suggestion
the `num_beams` greater than 1, and set `do_sample=True` to use this decoding strategy.
```

stevhliu commented on docs/source/en/generation_strategies.mdx:
````diff
@@ -0,0 +1,287 @@
+```
+
+### Diverse beam search decoding
+The Diverse beam search decoding strategy is an extension of the beam search strategy that allows to generate a more diverse
````
```suggestion
The diverse beam search decoding strategy is an extension of the beam search strategy that allows for generating a more diverse
```

stevhliu commented on docs/source/en/generation_strategies.mdx:
```diff
@@ -0,0 +1,287 @@
+### Diverse beam search decoding
+The Diverse beam search decoding strategy is an extension of the beam search strategy that allows to generate a more diverse
+set of beam sequences to choose from. To learn how it works, refer to [Diverse Beam Search: Decoding Diverse Solutions from Neural Sequence Models](https://arxiv.org/pdf/1610.02424.pdf).
+The Diverse Beam Search approach has two main parameters: `num_beams` and `num_beam_groups`.
```
```suggestion
This approach has two main parameters: `num_beams` and `num_beam_groups`.
```

stevhliu commented on docs/source/en/generation_strategies.mdx:
````diff
@@ -0,0 +1,287 @@
+'The Design Principles are a set of universal design principles that can be applied to any location, climate and culture, and they allow us to design the most efficient and sustainable human habitation and food production systems.'
+```
+
+This guide illustrates the main parameters that enable various decoding strategies. More advanced parameters exist for the
+generate() method, which give you even further control over the generate() method's behavior.
````
```suggestion
This guide illustrates the main parameters that enable various decoding strategies. More advanced parameters exist for the
[`generate`] method, which gives you even further control over the [`generate`] method's behavior.
```

stevhliu commented on docs/source/en/generation_strategies.mdx:
```diff
@@ -0,0 +1,287 @@
+For the complete list of the available parameters, refer to the [API documentation](./main_classes/text_generation.mdx).
+
+## Save a custom decoding strategy with your model
+If you would like to share your fine-tuned model with a specific generation configuration, you can do so:
```
```suggestion
If you would like to share your fine-tuned model with a specific generation configuration, you can:
```

stevhliu commented on docs/source/en/generation_strategies.mdx:
```diff
@@ -0,0 +1,287 @@
+
+## Save a custom decoding strategy with your model
+If you would like to share your fine-tuned model with a specific generation configuration, you can do so:
+* Create a `GenerationConfig` class instance
```
```suggestion
* Create a [`GenerationConfig`] class instance
```

stevhliu commented on docs/source/en/generation_strategies.mdx:
```diff
@@ -0,0 +1,287 @@
+If you would like to share your fine-tuned model with a specific generation configuration, you can do so:
+* Create a `GenerationConfig` class instance
+* Specify the decoding strategy parameters
+* Save your generation configuration with `save_pretrained`, making sure to leave its `config_file_name` argument empty
```
```suggestion
* Save your generation configuration with [`GenerationConfig.save_pretrained`], making sure to leave its `config_file_name` argument empty
```

stevhliu commented on docs/source/en/generation_strategies.mdx:
````diff
@@ -0,0 +1,287 @@
+>>> generation_config.save_pretrained("my_account/my_model", push_to_hub=True)
+```
+
+You can also store several generation parametrizations in a single directory, making use of the `config_file_name`
````
```suggestion
You can also store several generation parameters in a single directory, making use of the `config_file_name`
```

MKhalusova replied:
I think it should've been "several generation configurations".

stevhliu commented on docs/source/en/generation_strategies.mdx:
````diff
@@ -0,0 +1,287 @@
+```
+
+You can also store several generation parametrizations in a single directory, making use of the `config_file_name`
+argument in `save_pretrained`. You can later instantiate them with `from_pretrained`. This is useful if you want to
````
```suggestion
argument in [`GenerationConfig.save_pretrained`]. You can later instantiate them with [`GenerationConfig.from_pretrained`]. This is useful if you want to
```

stevhliu commented on docs/source/en/generation_strategies.mdx:
```diff
@@ -0,0 +1,287 @@
+You can also store several generation parametrizations in a single directory, making use of the `config_file_name`
+argument in `save_pretrained`. You can later instantiate them with `from_pretrained`. This is useful if you want to
+store several generation configurations for a single model (e.g. one for creative text generation with sampling, and
+other for summarization with beam search). You must have the right Hub permissions to add configuration files to a model.
```
```suggestion
one for summarization with beam search). You must have the right Hub permissions to add configuration files to a model.
```

stevhliu reviewed (approved):
Super nice and useful! 🚀 

I wonder if it'd make more sense to move the last section to after the "Default text generation configuration". I think having them next to each other could be nice since they're related topics. The question that naturally arises after seeing what the default is, is how can I customize it?

MKhalusova commented on docs/source/en/generation_strategies.mdx:
```diff
@@ -0,0 +1,287 @@
+  "bos_token_id": 50256,
+  "eos_token_id": 50256,
+  "transformers_version": "4.26.0.dev0"
+}
```
Fixed in 8ec3e7401b778bc9efedce240ac68db46bf62071

MKhalusova committed:
Apply Steven's suggestions from review

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

MKhalusova committed:
new lines after headers to fix link rendering

MKhalusova commented on docs/source/en/generation_strategies.mdx:
````diff
@@ -0,0 +1,287 @@
+}
+```
+
+If a model has any modified values in the generation configuration, you can review them at this point. Note, however,
````
I'll rephrase. I meant that when a config if printed out, only modified values are printed.

MKhalusova commented on docs/source/en/generation_strategies.mdx:
```diff
@@ -0,0 +1,287 @@
+Here, we'll show what parameters control some of the strategies, and illustrate how you can use them.
+
+### Greedy Search
+Greedy search is the default decoding strategy. For generate() method to use it, `num_beams` has to be set to 1, and
```
Wouldn't have guessed that, thanks!

MKhalusova committed:
review feedback addressed. added links to image captioning and audio transcription examples

MKhalusova committed:
minor capitalization fix

sgugger commented on docs/source/en/generation_strategies.mdx:
```diff
@@ -0,0 +1,306 @@
+* [Audio transcription](./model_doc/whisper#transformers.WhisperForConditionalGeneration.forward.example)
+
+Note that the inputs to the generate method depend on the model's modality:
+* `input_ids` for the decoder-only models
+* `input_ids`, `input_values`, `input_features`, or `pixel_values` for the encoder-decoder models
```
This is not in link with the previous sentence. You should say it's `input_ids` for text, `input_values` or `input_features` for speech and `pixel_values` for vision. Decoder-only model that can generate are necessarily text models, but I don't think this is the right way to explain this.

sgugger commented on docs/source/en/generation_strategies.mdx:
```diff
@@ -0,0 +1,306 @@
+Note that the inputs to the generate method depend on the model's modality:
+* `input_ids` for the decoder-only models
+* `input_ids`, `input_values`, `input_features`, or `pixel_values` for the encoder-decoder models
+In some cases, such as visual question answering, the inputs should have both `pixel_values` and `input_ids`.
```
Maybe tell the user to look at the documentation of their specific model in case of doubt? This seems pretty niche for this generic guide.

sgugger reviewed (approved):
Thanks for iterating! Left a couple of last comments and we should be good to go!

MKhalusova committed:
addressed the review feedback

gante commented on docs/source/en/generation_strategies.mdx:
```diff
@@ -0,0 +1,304 @@
+<!--Copyright 2020 The HuggingFace Team. All rights reserved.
```
```suggestion
<!--Copyright 2023 The HuggingFace Team. All rights reserved.
```

gante commented on docs/source/en/generation_strategies.mdx:
````diff
@@ -0,0 +1,304 @@
+You can override any `generation_config` by passing the parameters and their values directly to the [`generate`] method:
+
+```python
+>>> my_model.generate(inputs, num_beams=4, do_sample=True)
````
```suggestion
>>> my_model.generate(**inputs, num_beams=4, do_sample=True)
```

In all examples, I'm going to ask to change the first argument to `generate` from `inputs` (a single tensor) to `**inputs` (an unpacked dictionary of tensors), or similar.

This is because the tokenizer/preprocessor may prepare more than one input to the model -- for instance, in NLP models, the tokenizer produces at least the input tokens and the attention mask. While `generate` often works well without these secondary inputs, it has to make assumptions about them (and fails sometimes). Since we are writing these amazing docs from scratch, we can take the opportunity to guide the users towards a better use of generate 🤗 

I've double-checked the stand-alone examples and confirmed that they work after the change.

MKhalusova replied:
Thanks a ton for the detailed explanation! I appreciate it!

gante commented on docs/source/en/generation_strategies.mdx:
```diff
@@ -0,0 +1,304 @@
+>>> checkpoint = "distilgpt2"
+
+>>> tokenizer = AutoTokenizer.from_pretrained(checkpoint)
+>>> input_ids = tokenizer(prompt, return_tensors="pt").input_ids
+
+>>> model = AutoModelForCausalLM.from_pretrained(checkpoint)
+>>> outputs = model.generate(input_ids)
```
```suggestion
>>> inputs = tokenizer(prompt, return_tensors="pt")

>>> model = AutoModelForCausalLM.from_pretrained(checkpoint)
>>> outputs = model.generate(**inputs)
```

gante commented on docs/source/en/generation_strategies.mdx:
```diff
@@ -0,0 +1,304 @@
+>>> model = AutoModelForCausalLM.from_pretrained(checkpoint)
+
+>>> prompt = "Hugging Face Company is"
+>>> input_ids = tokenizer(prompt, return_tensors="pt").input_ids
+
+>>> outputs = model.generate(input_ids, penalty_alpha=0.6, top_k=4, max_new_tokens=100)
```
```suggestion
>>> inputs = tokenizer(prompt, return_tensors="pt")

>>> outputs = model.generate(**inputs, penalty_alpha=0.6, top_k=4, max_new_tokens=100)
```

gante commented on docs/source/en/generation_strategies.mdx:
```diff
@@ -0,0 +1,304 @@
+>>> model = AutoModelForCausalLM.from_pretrained(checkpoint)
+
+>>> prompt = "Today was an amazing day because"
+>>> input_ids = tokenizer(prompt, return_tensors="pt").input_ids
+
+>>> outputs = model.generate(input_ids, do_sample=True, max_new_tokens=100)
```
```suggestion
>>> inputs = tokenizer(prompt, return_tensors="pt")

>>> outputs = model.generate(**inputs, do_sample=True, max_new_tokens=100)
```

gante commented on docs/source/en/generation_strategies.mdx:
```diff
@@ -0,0 +1,304 @@
+>>> checkpoint = "gpt2-medium"
+
+>>> tokenizer = AutoTokenizer.from_pretrained(checkpoint)
+>>> input_ids = tokenizer(prompt, return_tensors="pt").input_ids
+
+>>> model = AutoModelForCausalLM.from_pretrained(checkpoint)
+
+>>> outputs = model.generate(input_ids, num_beams=5, max_new_tokens=50)
```
```suggestion
>>> inputs = tokenizer(prompt, return_tensors="pt")

>>> model = AutoModelForCausalLM.from_pretrained(checkpoint)

>>> outputs = model.generate(**inputs, num_beams=5, max_new_tokens=50)
```

gante commented on docs/source/en/generation_strategies.mdx:
```diff
@@ -0,0 +1,304 @@
+>>> checkpoint = "t5-small"
+
+>>> tokenizer = AutoTokenizer.from_pretrained(checkpoint)
+>>> input_ids = tokenizer(prompt, return_tensors="pt").input_ids
+
+>>> model = AutoModelForSeq2SeqLM.from_pretrained(checkpoint)
+
+>>> outputs = model.generate(input_ids, num_beams=5, do_sample=True)
```
```suggestion
>>> inputs = tokenizer(prompt, return_tensors="pt")

>>> model = AutoModelForSeq2SeqLM.from_pretrained(checkpoint)

>>> outputs = model.generate(**inputs, num_beams=5, do_sample=True)
```

gante commented on docs/source/en/generation_strategies.mdx:
```diff
@@ -0,0 +1,304 @@
+>>> efficient way possible."
+
+>>> tokenizer = AutoTokenizer.from_pretrained(checkpoint)
+>>> input_ids = tokenizer(prompt, return_tensors="pt").input_ids
+
+>>> model = AutoModelForSeq2SeqLM.from_pretrained(checkpoint)
+
+>>> outputs = model.generate(input_ids, num_beams=5, num_beam_groups=5, max_new_tokens=30)
```
```suggestion
>>> inputs = tokenizer(prompt, return_tensors="pt")

>>> model = AutoModelForSeq2SeqLM.from_pretrained(checkpoint)

>>> outputs = model.generate(**inputs, num_beams=5, num_beam_groups=5, max_new_tokens=30)
```

gante commented on docs/source/en/generation_strategies.mdx:
```diff
@@ -0,0 +1,304 @@
+* [Audio transcription](./model_doc/whisper#transformers.WhisperForConditionalGeneration.forward.example)
+
+Note that the inputs to the generate method depend on the model's modality. They can be `input_ids` for text, `input_values`
+or `input_features` for speech and `pixel_values` for computer vision. When in doubt, refer to the model's documentation.
```
I would add here a sentence related to the other suggested changes I've made. 

Something like: "Please note that our models' preprocessor classes (such as the `AutoTokenizer` or the `AutoProcessor`) may return more than one model input when called, in which case you should feed them all to `generate()`".

MKhalusova replied:
Modified in 50146bfd463d4d8bd94887658329b208c2f0e30f

gante reviewed (approved):
Thank you for these great docs, I think users will greatly appreciate them <3

I left a few nits, most of them were about adopting better practices in the examples (practices that were poorly documented before!)

gante commented on docs/source/en/generation_strategies.mdx:
```diff
@@ -0,0 +1,304 @@
+[`generate`] method, which gives you even further control over the [`generate`] method's behavior.
+For the complete list of the available parameters, refer to the [API documentation](./main_classes/text_generation.mdx).
+
+## Save a custom decoding strategy with your model
```
Do you think it would make sense to move this to before "## Decoding strategies"? i.e. shape the docs to first show how to manipulate generate in general, then dive into its options.

I'm concerned users will miss this new functionality (because it's at the bottom), especially with the example removal in #21112

MKhalusova replied:
Sure, folks seem to agree that your suggested order would be better. I moved it up, before the ## decoding strategies, in 50146bfd463d4d8bd94887658329b208c2f0e30f

MKhalusova committed:
Apply suggestions from review

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

MKhalusova committed:
Applied review suggestions

MKhalusova commented:
Hi! Thank you all for the invaluable feedback! I have addressed all the comments and suggestions. Please let me know if there's anything else I need to improve or if we can merge this PR and https://github.com/huggingface/transformers/pull/21112.

gante commented:
I think both can be merged. Thank you @MKhalusova ❤️
